### PR TITLE
Stasis Bags

### PR DESCRIPTION
**Pull request by jakmak6**
See https://github.com/vgstation-coders/vgstation13/issues/15561

I have resurrected stasis bags/cryobags from our code. These bags are one use bodybags that if you put someone inside of it stops all bleeding/oxyloss/toxin buildup/chem metabolism/ floor dragging harm/ ect.

The bags in turn for being inside them give a slow but steady uptick in brain damage so you cant keep someone in stasis forever.

However there are 2 types of stasis bags. A simple one that does the brain damage and a advanced one that does no brain damage while loosing none of the advantages.

Currently no way to get these bags other than adminbus. I will add in some to paramedic locker/protolathe so until then DMN

Credit to baystation for the advanced cryobag sprite

### CODE REVIEW

**Comment by CptWad** [comment]
jakmak and his clone commit army

**Comment by CptWad** [comment]
> I will add in some to paramedic locker/protolathe so until then DMN

pls dont use the map, just add it directly to the paramedic closet content

**Comment by jakmak6** [comment]
i intend to.

**Comment by Exxion** [comment]
Comment or something when you want the DNM taken off

**Comment by jakmak6** [comment]
ye i will do that

So deciding how to get them. This is what i want. For the standard brain damage bags cargo order locked in paramed access crate, box of 4-5 in parameds locker, A few in vox skipjack, protolathable for 3 sheets iron, 1 sheet glass, 3 sheets plastic, and 3 sheets of plasma with research levels plasma 2, bio 4, material 2, and bluespace 3

Advanced bag is only gettable by protolathe, ERT base, Deathsquad base, nukeop ship (maybe?) Protolathe Material price 3 sheets of iron, 1 sheet of glass, 3 sheets of plastic, 5 sheets of plasma ,2 sheets of gold, 1 sheet of uranium Science level needs to be plasma 4, bio 5, material 3, and blue space 4.

**Comment by jakmak6** [comment]
Maybe i could raise the brain damage rate on the shitty bags. .2 per tick is a bit low

**Comment by CptWad** [comment]
Cool.
Why plastic?

**Comment by jakmak6** [comment]
because what else would a body bag be made out of?

**Comment by CptWad** [comment]
neometallic fiberous material

**Comment by jakmak6** [comment]
eh your probably right. ill lower the plastic cost and raise the metal cost

File: code/game/objects/items/bodybag.dm [hunk 1]
```diff
@@ -39,7 +39,8 @@
 	icon_closed = "bodybag_closed"
 	icon_opened = "bodybag_open"
 	density = 0
-
+	var/cryobag_reg = 0
```
Comment (PJB3005) [inline code/game/objects/items/bodybag.dm:42]: Nope. Not happening. Use inheritance properly.

File: code/game/objects/items/bodybag.dm [hunk 2]
```diff
@@ -113,22 +123,103 @@
 	icon_closed = "bodybag_closed"
 	icon_opened = "bodybag_open"
 	density = 0
-
 	var/used = 0
+	cryobag_reg = 1
+
+/*/obj/structure/closet/body_bag/cryobag/MouseDrop(over_object, src_location, over_location)
```
Comment (PJB3005) [inline code/game/objects/items/bodybag.dm:129]: Remove this.

File: code/game/objects/items/bodybag.dm [hunk 3]
```diff
@@ -113,22 +123,103 @@
...
+	. = ..()
+	if(used)
+		var/obj/item/O = new/obj/item(src.loc)
+		O.name = "used advanced stasis bag"
+		O.icon = src.icon
+		O.icon_state = "bodybag_used"
+		O.desc = "Pretty useless now.."
+		qdel(src)
+
+/obj/item/bodybag/cryobag/advanced/attack_self(mob/user)
```
Comment (PJB3005) [inline code/game/objects/items/bodybag.dm:199]: Reorder this so it's below the main item definition.

File: code/game/objects/items/bodybag.dm [hunk 4]
```diff
@@ -113,22 +123,103 @@
...
+
+	var/no_brain_damage = 1 // so the adv bag doesnt brain damage
+	cryobag_adv = 1
+	cryobag_reg = 0
+
+
+/obj/structure/closet/body_bag/cryobag/advanced/open()
+	. = ..()
+	if(used)
+		var/obj/item/O = new/obj/item(src.loc)
```
Comment (PJB3005) [inline code/game/objects/items/bodybag.dm:192]: Using a bare `/item` is a terrible idea. Make a subtype.

File: code/modules/mob/living/carbon/human/life.dm [hunk 5]
```diff
@@ -80,6 +80,7 @@ var/global/list/organ_damage_overlays = list(
 	var/prev_gender = null // Debug for plural genders
 	var/temperature_alert = 0
 	var/in_stasis = 0
+	var/adv_stasis = 0 //for the adv cryobag
```
Comment (PJB3005) [inline code/modules/mob/living/carbon/human/life.dm:83]: Nope. This shouldn't be tracked on the mob.

File: code/modules/mob/living/carbon/human/life.dm [hunk 6]
```diff
@@ -128,6 +129,7 @@ var/global/list/organ_damage_overlays = list(
 
 	var/datum/gas_mixture/environment = loc.return_air()
 	in_stasis = istype(loc, /obj/structure/closet/body_bag/cryobag) && loc:opened == 0 //Nice runtime operator
+	adv_stasis = istype(loc, /obj/structure/closet/body_bag/cryobag/advanced) && loc:opened == 0
```
Comment (PJB3005) [inline code/modules/mob/living/carbon/human/life.dm:132]: Nice runtime operator.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Use my implementation please
https://github.com/ParadiseSS13/Paradise/pull/7903

**Comment by Moonman1946** [comment]
Being a good paramedic means stabilizing someone before bringing him to medbay, and carrying a roller bed or something so he doesn't lose blood on the way there.
The bag completely deletes the need for both, changing the paramedic into a click&drag bot, might as well hire an assistant, give him bunch of bags and nothing else and tell him to do it.

**Comment by ghost** [comment]
If the person inside can't escape the stasis bag it is a great murder item, push somebody down and very quickly drop them into a stasis bag. You could space the bag and it's hard to stop. You could say the same for the parapen but dragging a regular body is noticeable, somebody dragging a stasis bag probably wouldn't be noticed, and nobody is going to try and open it like they do with lockers/bodybags.

It's a cool idea but I think even the basic one should be locked behind r&d, cost quite a bit in asteroid materials and there shouldn't be any roundstart. Except maybe the vox trader ship has a chance to spawn with a basic one, as something to sell.

**Comment by jakmak6** [comment]
Moon you need to remember the average paramedic isn't that smart. The roller bed when deployed sucks to have to work with in compact spaces and this item being locked behind cmo/paramedic access roundstart means that just like the biofoam it will be seldom used. Sure a good doc would go through the effort of procuring them but just like the biofoam most won't use them.

And donnager you can leave the stasis bag ad anytime provided you are not incapacitated or stunned.  The bag does not put you in some form of a permanent stun. It's like being put in a unlocked locker. You can walk away at any time

I don't know now if I want to give some to the paramedic locker now. I will think about it. For now they won't be in the paramedic locker

**Comment by ghost** [comment]
I thought the bag froze you up kinda like cryo, if it's just a bag you can leave that's fine then.

I still think it should have to be bought or made in research, not something for free.

**Comment by jakmak6** [comment]
Yeah I can see where the confusion would come from with that. I want to add new RnD stuff to make sci med department cooperation less of a shitty meme and more of a decent quality meme but at the same time paramedic could use more goodies to make the job more unique and attractive instead of just going normal doc. It's partially why I went with 2 separate bags. A shitty one for low Matt cost and for roundstart paramedic and cargo and a high quality one that's mineral expensive, no down side and protolathe only. But I am still taking suggestions

**Comment by ghost** [comment]
Your code doesn't solve problems with stasis bags. They halting only certain effects looping in human Life() while a lot of bleeding/poison/dying processes stored in a mob Life().

Also you're doing same thing oldcoders did with bags. Fucking bag loc check right in the Life() proc. Why do you need this, again? Can't you figure out a better approach for it? Why you need to call istype on current loc each 2 seconds?

My proposals:

- Look at my code i linked above
- Remove loc check from Life() and use define from my code above instead
- Make stasis halt mob Life() so it may save patient from everything but cloneloss and braindamage
- Give bag ability to be lockable by medical ID

**Comment by Shadowmech88** [comment]
@jakmak6 do you plan to continue working on this?

**Comment by PJB3005** [comment]
Considering this is jakmak and it's been dead for months, I'm closing this. Comment if you want to have it reopened.

**Pull request closed without merging**